kernel/src/table_changes/log_replay/tests.rs

```diff
@@ -0,0 +1,609 @@
+    end_version: impl Into<Option<Version>>,
+) -> DeltaResult<Vec<ParsedLogPath>> {
+    let table_root = url::Url::from_directory_path(path).unwrap();
+    let log_root = table_root.join("_delta_log/")?;
+    let log_segment = LogSegment::for_table_changes(
+        engine.get_file_system_client().as_ref(),
+        log_root,
+        start_version,
+        end_version,
+    )?;
+    Ok(log_segment.ascending_commit_files)
+}
+
+fn result_to_sv(iter: impl Iterator<Item = DeltaResult<TableChangesScanData>>) -> Vec<bool> {
+    iter.map_ok(|scan_data| scan_data.selection_vector.into_iter())
+        .flatten_ok()
+        .try_collect()
+        .unwrap()
+}
+
+#[tokio::test]
+async fn metadata_protocol() {
+    let engine = Arc::new(SyncEngine::new());
+    let mut mock_table = LocalMockTable::new();
+    let schema_string = serde_json::to_string(&get_schema()).unwrap();
+    mock_table
+        .commit([
+            Action::Metadata(Metadata {
+                schema_string,
+                configuration: HashMap::from([
+                    ("delta.enableChangeDataFeed".to_string(), "true".to_string()),
+                    (
+                        "delta.enableDeletionVectors".to_string(),
+                        "true".to_string(),
+                    ),
+                    ("delta.columnMapping.mode".to_string(), "none".to_string()),
+                ]),
+                ..Default::default()
+            }),
+            Action::Protocol(
+                Protocol::try_new(
+                    3,
+                    7,
+                    Some([ReaderFeatures::DeletionVectors]),
+                    Some([ReaderFeatures::ColumnMapping]),
+                )
+                .unwrap(),
+            ),
+        ])
+        .await;
+
+    let commits = get_segment(engine.as_ref(), mock_table.table_root(), 0, None)
+        .unwrap()
+        .into_iter();
+
+    let scan_batches =
+        table_changes_action_iter(engine, commits, get_schema().into(), None).unwrap();
+    let sv = result_to_sv(scan_batches);
+    assert_eq!(sv, &[false, false]);
+}
+#[tokio::test]
+async fn cdf_not_enabled() {
+    let engine = Arc::new(SyncEngine::new());
+    let mut mock_table = LocalMockTable::new();
+    let schema_string = serde_json::to_string(&get_schema()).unwrap();
+    mock_table
+        .commit([Action::Metadata(Metadata {
+            schema_string,
+            configuration: HashMap::from([(
+                "delta.enableDeletionVectors".to_string(),
+                "true".to_string(),
+            )]),
+            ..Default::default()
+        })])
+        .await;
+
+    let commits = get_segment(engine.as_ref(), mock_table.table_root(), 0, None)
+        .unwrap()
+        .into_iter();
+
+    let res: DeltaResult<Vec<_>> =
+        table_changes_action_iter(engine, commits, get_schema().into(), None)
+            .and_then(|iter| iter.try_collect());
+
+    assert!(matches!(res, Err(Error::ChangeDataFeedUnsupported(_))));
```

Comment:
For unit tests, we should probably know whether the "outer" result is an Err, or if the iterator returns some "inner" Err. 

Here for example, I _think_ the function returns Err directly and we can just match on it:
```suggestion
    let res = table_changes_action_iter(engine, commits, get_schema().into(), None);
    assert!(matches!(res, Err(Error::ChangeDataFeedUnsupported(_))));
```

Otherwise, we should `unwrap` the function result and then attempt the `try_collect` directly. Or could even do something like
```suggestion
    let res = table_changes_action_iter(engine, commits, get_schema().into(), None)
            .unwrap()
            .find(Result::is_err)
    assert!(matches!(res, Some(Err(Error::ChangeDataFeedUnsupported(_)))));
```

(several more below)